Add Fundle install support

SirWrexes opened the pull request:
Hey there,

I'm not a fan of installing plugins straight into where I put my own configs.
Nonetheless, I've been using Fisher for a while until, recently, jorgebucaran/fisher#651 broke everything.
I kept getting annoying warnings telling me to delete my plugin files in order to update them blablabla... but I digress.

I switched to Fundle, which works more elegantly IMO by placing plugins in a dedicated directory tree then sourcing everything.

Luckily, there are several ways of checking if a user has Fundle, and one of them is querying `functions` too see if `__fundle_plugins_dir` exists, which will then serve to tell us where Gitnow has been installed.
I quickly wrote a fix with this method, while retaining the pretty handy variables you created.

Note that `fish_config` could also be set using the value from the standard `__fish_config_dir` variable.

Have a good day ! :wave:

joseluisq committed:
refactor: improve performance of all regex-based system commands

this commit also speeds up gitnow commands like:
tag, logs, move, gitflow, github and bitbucket

joseluisq committed:
2.5.1

joseluisq committed:
feat: paket support

joseluisq committed:
refactor: simplify paket install and uninstall events

joseluisq committed:
refactor: relocate gitnow config file

joseluisq committed:
docs: minor keybinding descriptions

jorgebucaran committed:
Upgrade Fisher instructions

jorgebucaran committed:
Fix CI

lukasgierth committed:
add rudimantary merge command with simple branch completion

lukasgierth committed:
add manual entry (without keyboard shortcut)

lukasgierth committed:
do some checks, add abort option to merge, update README

joseluisq committed:
fix: invalid argument on check using __gitnow_check_if_branch_exist

lukasgierth committed:
fix naming, merge not move

lukasgierth committed:
add continue as option, capture unsupported --options

lukasgierth committed:
fix echo message

joseluisq committed:
refactor: paket support on ci

joseluisq committed:
refactor: print package manager versions

joseluisq committed:
2.6.0

joseluisq committed:
feat: fatal matched multiple remote tracking branches on move command

joseluisq committed:
feat: project sponsor

joseluisq committed:
feat: github actions

SirWrexes committed:
fix: Use -query when calling type

--quiet is deprecated (https://github.com/fish-shell/fish-shell/commit/bfb5b28d0f69fe499817d21b1811c3083f76445d)

SirWrexes committed:
fix: Use -query when calling type

--quiet is deprecated (https://github.com/fish-shell/fish-shell/commit/bfb5b28d0f69fe499817d21b1811c3083f76445d)

SirWrexes committed:
fix: Use -q instead of --query for compatiblity

SirWrexes committed:
fix: Use -q instead of --query for compatiblity

joseluisq committed:
test: fish files error checks

joseluisq committed:
test: another error string check

joseluisq committed:
refactor: config file parsing to increase performance startup 62% avg

Config file parsing (.gitnow) was written completely and as result it
boosts performance startup of GitNow up to 62% faster avg.

Stats:

Old parser (v2.6.0):
Executed in  223,04 millis    fish           external
   usr time  209,69 millis  636,00 micros  209,06 millis
   sys time   31,97 millis  175,00 micros   31,80 millis

New parser:
Executed in   84,62 millis    fish           external
   usr time   63,87 millis    0,00 micros   63,87 millis
   sys time   24,04 millis  921,00 micros   23,12 millis

joseluisq committed:
refactor: improve performance startup of logs command

it limits max number of commits to 80 by default

joseluisq committed:
2.7.0

joseluisq committed:
refactor: paket file events definition

SirWrexes committed:
feat: Add Fundle install support

joseluisq commented:
> I switched to Fundle, which works more elegantly IMO by placing plugins in a dedicated directory tree then sourcing everything.
> 
> Luckily, there are several ways of checking if a user has Fundle, and one of them is querying `functions` too see if `__fundle_plugins_dir` exists, which will then serve to tell us where Gitnow has been installed.
> I quickly wrote a fix with this method, while retaining the pretty handy variables you created.

IMO your PR looks more like a feature rather than a fix since as I understood you want to add `Fundle` support. Right?
If so please update the title to "Add Fundle install support" or something.

> Note that `fish_config` could also be set using the value from the standard `__fish_config_dir` variable.

Yeah, you could send a new PR introducing that change if you want. `$XDG_CONFIG_HOME` is kind of restricted.

joseluisq commented on conf.d/gitnow_config.fish:
```diff
@@ -8,11 +8,19 @@ set -q fish_functions; or set -g fish_functions "$fish_config/functions"
 set -q GITNOW_CONFIG_FILE; or set -g GITNOW_CONFIG_FILE ~/.gitnow
 
-set -g gitnow_version 2.6.0
+set -g gitnow_version 2.7.0
```
Why that change? Are you using the latest master? Can you sync your PR branch please?
https://github.com/joseluisq/gitnow/blob/master/conf.d/gitnow_config.fish#L11

SirWrexes replied:
Oh my bad. I simply pulled from master when forking, then made the `fundle-compat` branch after editting the file.

SirWrexes replied:
Wait, on master it's 2.6 too.
I honestly don't know what happened there, that is quite odd. I'll revert it.

joseluisq replied:
Basically you need to sync your fork first.

joseluisq reviewed (commented):
It would be also great if you can share the steps to test Gitnow with Fundle.

joseluisq commented:
And probably you also want to update the `Install` section of `README` with the new install method. https://github.com/joseluisq/gitnow#install

SirWrexes commented:
I'm working on adding the other things like editing the README etc.
So, what version number should I keep ? 2.7.0 I assume ?

joseluisq commented:
> So, what version number should I keep ? 2.7.0 I assume ?

Yes, Just let it as it is as current Gitnow `master`.

SirWrexes commented:
Alright.

I have added the installation instructions to `README.md`, and squashed that plus the edits to `gitnow_config.fish` into one `feat:` commit.

I also took time to add a disclaimer about 2.7.0 not supporting Fundle yet in the "Latest stable" installation instructions. Since Fundle's README isn't very clear about how to get a particular version of a plugin and it's not extremely user friendly, I thought it was better having the installation method + a note rather than nothing. It will save you some time later too I guess.

Synchronising my fork in between two commits made squashing take longer than expected but everything seems to be in order and working fine.
I tried it out by copying my fork checked out on `fundle-compat` in my Fundle plugin dir and nothing weird came up.

I'll push the squashed commits.

SirWrexes committed:
Merge branch 'master' into fundle-compat

SirWrexes commented:
Wow seeing the whole history come up in there because of the sync issue sure is a scary thing.
My apologies. I'm kinda new to the PR thing.

joseluisq commented:
> Wow seeing the whole history come up in there because of the sync issue sure is a scary thing.
> My apologies. I'm kinda new to the PR thing.

Understand, but unfortunately the history is a bit messy I recommend do a new PR with just your commit changes.

joseluisq commented:
Feel free to sync your fork or re-fork Gitnow again and open a new PR again.

SirWrexes commented:
I totally understand.
Will do !
